core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
 
     brokerTopicStats.updateReplicationBytesIn(records.sizeInBytes)
 
+    val highWatermarkChanged = log.maybeUpdateHighWatermark(partitionData.highWatermark)
+    if (highWatermarkChanged) {
+      logAppendInfo.foreach { _ => partitionsWithNewHighWatermark += topicPartition }
+      if (logTrace)
+        trace(s"Follower updated replica high watermark for partition $topicPartition to ${partitionData.highWatermark}")
```

Comment:
@dajac 

this statement is not entirely true and this was the reason i've attempted to use the HighWatermarkUpdated case class.  since this log is important, i think we should still have a case class for this.

Reply:
I see. How about returning an `Option` in `maybeUpdateHighWatermark` which would contain the updated HWM only if it was changed. Would this work?

I would also update the trace message to include both the received HWM from the leader and the updated HWM. Both seems useful here.

Reply:
much more elegant. thanks for the suggestion!